File: js/src/models/contracts.test.ts
// The shipped clients' CONTRACT DECLARATION, checked in the test suite too.
//
// The real check is the static assertion at the bottom of each client module —
// `tsc --noEmit` fails there if a client drifts from the published shape. This
// file makes the claim visible in `npm test` as well, and exercises it through
// the contract rather than through the concrete class, which is the thing that
// actually matters to a caller.

import { describe, expect, expectTypeOf, it } from "vitest";

import type { EmbeddingProvider, GeneratorProvider } from "../contracts.js";
import { AnthropicGenerator } from "./anthropic.js";
import { OpenAIEmbedder } from "./embed.js";
import { OpenAIChatGenerator, type Transport } from "./openai.js";

describe("the shipped clients satisfy the published contracts", () => {
  it("OpenAIEmbedder is an EmbeddingProvider", () => {
    expectTypeOf<OpenAIEmbedder>().toMatchTypeOf<EmbeddingProvider>();
  });

  it("both generators are one GeneratorProvider", () => {
    // An Anthropic-shaped endpoint and an OpenAI-shaped one are
    // indistinguishable to a caller — the whole point of the contract.
    expectTypeOf<OpenAIChatGenerator>().toMatchTypeOf<GeneratorProvider>();
    expectTypeOf<AnthropicGenerator>().toMatchTypeOf<GeneratorProvider>();
  });

  // R3, checked on a real client: the transport lives in the CONSTRUCTOR, so
  // the contract a provider implements says nothing about HTTP. A provider that
  // never opens a socket satisfies exactly the same interface this client does.
  it("the contract says nothing about the transport", async () => {
    const transport: Transport = () =>
      JSON.stringify({ choices: [{ message: { content: "the passage" } }] });
    const provider: GeneratorProvider = new OpenAIChatGenerator(
      { base_url: "http://example.invalid/v1", model: "qwen" },
      transport,
    );
    expect(await provider.answer("q", "the passage", "en")).toBe("the passage");
  });

  it("the embedding client answers the batch contract, in input order", async () => {
    const transport: Transport = () =>
      JSON.stringify({ data: [{ embedding: [1, 0] }, { embedding: [0, 1] }] });
    const provider: EmbeddingProvider = new OpenAIEmbedder(
      { base_url: "http://example.invalid/v1", model: "bge-m3" },
      transport,
    );
    expect(await provider.embed(["a", "b"])).toEqual([
      [1, 0],
      [0, 1],
    ]);
  });
});

File: golang/models/contracts_test.go
package models

import (
	"testing"

	"github.com/muthuishere/citenexus/golang/contracts"
)

// The compile-time assertions in contracts.go are the real check — this test
// exists so the declaration is visible in the test output too, and so a reader
// scanning `go test -v` can see WHICH client claims WHICH seam.
func TestShippedClientsSatisfyThePublishedContracts(t *testing.T) {
	var (
		embedding = NewOpenAIEmbedding("http://x/v1", "bge-m3", nil)
		openai    = NewOpenAIChatGenerator("http://x/v1", "qwen", 0, nil, nil)
		anthropic = NewAnthropicGenerator("", "claude", 0, DefaultAnthropicMaxTokens, nil)
	)

	if _, ok := any(embedding).(contracts.EmbeddingProvider); !ok {
		t.Error("OpenAIEmbedding does not satisfy contracts.EmbeddingProvider")
	}
	for name, client := range map[string]any{
		"OpenAIChatGenerator": openai,
		"AnthropicGenerator":  anthropic,
	} {
		if _, ok := client.(contracts.GeneratorProvider); !ok {
			t.Errorf("%s does not satisfy contracts.GeneratorProvider", name)
		}
	}

	// The batch client also plugs into the deprecated single-text ingest seam
	// through the published adapter, with no glue written by the caller.
	if _, ok := any(contracts.SingleFrom(embedding)).(contracts.SingleTextEmbedder); !ok {
		t.Error("contracts.SingleFrom(OpenAIEmbedding) is not a SingleTextEmbedder")
	}
}

// R3, checked on a real client: the transport lives in the CONSTRUCTOR, so the
// contract a provider implements says nothing about HTTP. A provider that never
// opens a socket satisfies exactly the same interface these clients do.
func TestTheContractSaysNothingAboutTheTransport(t *testing.T) {
	var provider contracts.GeneratorProvider = NewOpenAIChatGenerator(
		"http://example.invalid/v1", "qwen", 0, nil,
		func(string, []byte, map[string]string) ([]byte, error) {
			return []byte(`{"choices":[{"message":{"content":"the passage"}}]}`), nil
		},
	)
	got, err := provider.Answer("q", "the passage", "en")
	if err != nil || got != "the passage" {
		t.Fatalf("Answer through the contract: %q %v", got, err)
	}
}
